sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuery.java

```diff
@@ -1016,6 +1017,42 @@ public GroupByQuery toGroupByQuery()
         grouping.getSubtotals().toSubtotalsSpec(grouping.getDimensionSpecs()),
         ImmutableSortedMap.copyOf(plannerContext.getQueryContext())
     );
+    // in this case, the timestampResult optimization is not neccessary
```

Comment:
There are many reasons let me think this optimization unnecessary for this case.

When `query.isApplyLimitPushDown()` is true, it means the limit is calculated on compute node first.

`select floor(__time to hour),dim1 from a_table group by floor(__time to hour),dim1 limit 10`
On compute node,  the row number produced with limit when granularity=ALL is different with the sum of row number produced with limit in each hour.
The `limit` should be applied to the whole result rows.

`select dim1,floor(__time to hour),dim2 from a_table group by dim1,floor(__time to hour),dim2 limit 10`
For this even more complicated case, if we replace granularity=ALL with granularity=HOUR, we have to modify the Buffer Grouper logic to adjust the time_floor dimension value(and the position in row) in the result row to make it looks the same with the result row when granularity=ALL, and not only the value but also the sequence in rows after the `limit` is applied. I don't think it is worth for both logic complexity or performance gain.

And for another reason, when the `limit` is pushed down to compute node, I think the whole computation should be fast enough, the timestampResult optimization is not going to be that effectively improve the total performance.

And the last reason, if we don't pushdown the `limit` to compute node and apply this optimization and do limit on broker, I guess the total performance is degraded.

Generally, the timestampResult optimization logic happens on broker, so I disable this optimization when `query.isApplyLimitPushDown()` is true

Reply:
Hmm sorry, I don't quite understand. I'm not sure what the limit clause does anything with this optimization.

> When `query.isApplyLimitPushDown()` is true, it means the limit is calculated on compute node first.

Yes, when `isApplyLimitPushDown` is true, the limit is applied in both the data nodes and the broker. In data nodes, the limit is applied when per-segment results are merged. The broker applies the limit again when it merges per-data-node results. For example, let's say you issued a groupBy query with a limit of 10, each data node would return at most 10 rows and the broker would merge those per-data-node results and apply the limit again to finally return only 10 rows.

> `select floor(__time to hour),dim1 from a_table group by floor(__time to hour),dim1 limit 10`
> On compute node, the row number produced with limit when granularity=ALL is different with the sum of row number produced with limit in each hour.
> The `limit` should be applied to the whole result rows.

I don't think the number of rows should be different in those cases. This optimization not just modifies the granularity but also dimensions. When granularity is set to ALL, the dimensions should have the virtual column for the timefloor function. When granularity is set to HOUR, the dimensions should not have that virtual column. In these cases, the actual keys used for grouping are effectively the same because the ALL granularity groups all rows into one bucket. As a result, the number of rows must be the same in both cases.

> `select dim1,floor(__time to hour),dim2 from a_table group by dim1,floor(__time to hour),dim2 limit 10`
> For this even more complicated case, if we replace granularity=ALL with granularity=HOUR, we have to modify the Buffer Grouper logic to adjust the time_floor dimension value(and the position in row) in the result row to make it looks the same with the result row when granularity=ALL, and not only the value but also the sequence in rows after the `limit` is applied. I don't think it is worth for both logic complexity or performance gain.

What logic should we modify? The optimization you added is modifying the native query that is passed all the way down to the data nodes which use the `BufferGrouper`. Why does pushing down the limit make any difference in the result signature of the data nodes?

> And for another reason, when the `limit` is pushed down to compute node, I think the whole computation should be fast enough, the timestampResult optimization is not going to be that effectively improve the total performance.

There is one performance issue with `LimitedBufferHashGrouper` that I'm aware of that makes the groupBy with limitPushDown on even slower than the same query with limitPushDown off. But even if there was no known performance issue, I think query performance is really different case by case and thus I wouldn't say that the query processing with limitPushDown is fast in data nodes and thus we need no more optimization.

> And the last reason, if we don't pushdown the `limit` to compute node and apply this optimization and do limit on broker, I guess the total performance is degraded.

Hmm, I'm confused. Are you saying that we should push down the limit because the performance will be worse otherwise? If so, I agree.

> Generally, the timestampResult optimization logic happens on broker, so I disable this optimization when `query.isApplyLimitPushDown()` is true

What do you mean by "the timestampResult optimization logic happens on broker"? The rewritten query will be passed down to the data nodes as far as I can tell.

Just for clarification, my question is not why we don't optimize this case now, but is whether your comment here is correct.